psl-core/src/main/java/org/linqs/psl/reasoner/sgd/term/SGDObjectiveTerm.java

```diff
@@ -96,40 +95,109 @@ public float evaluate(float[] variableValues) {
      */
-    public float minimize(int iteration, VariableTermStore termStore) {
+    public float minimize(int iteration, VariableTermStore termStore, float learningRate,
+            Map<Integer, Float> accumulatedGradientSquares,
+            Map<Integer, Float> accumulatedGradientMean,
+            Map<Integer, Float> accumulatedGradientVariance,
+            SGDExtension sgdExtension, boolean coordinateStep) {
         float movement = 0.0f;
-        float weight = rule.getWeight();
+        float variableStep = 0.0f;
+        float newValue = 0.0f;
+        float partial = 0.0f;
 
         GroundAtom[] variableAtoms = termStore.getVariableAtoms();
         float[] variableValues = termStore.getVariableValues();
+        float dot = dot(variableValues);
 
         for (int i = 0 ; i < size; i++) {
             if (variableAtoms[variableIndexes[i]] instanceof ObservedAtom) {
                 continue;
             }
 
-            float dot = dot(variableValues);
-            float gradient = computeGradient(i, dot);
-            float gradientStep = weight * gradient * (learningRate / iteration);
+            partial = computePartial(i, dot, rule.getWeight());
+            variableStep = computeVariableStep(variableIndexes[i], iteration, learningRate, partial,
+                    accumulatedGradientSquares, accumulatedGradientMean, accumulatedGradientVariance,
+                    sgdExtension);
 
-            float newValue = Math.max(0.0f, Math.min(1.0f, variableValues[variableIndexes[i]] - gradientStep));
+            newValue = Math.max(0.0f, Math.min(1.0f, variableValues[variableIndexes[i]] - variableStep));
             movement += Math.abs(newValue - variableValues[variableIndexes[i]]);
             variableValues[variableIndexes[i]] = newValue;
+
+            if (coordinateStep) {
+                dot = dot(variableValues);
+            }
         }
 
         return movement;
     }
 
-    private float computeGradient(int varId, float dot) {
+    private float computeVariableStep(
+            int variableIndex, int iteration, float learningRate, float partial,
+            Map<Integer, Float> accumulatedGradientSquares,
+            Map<Integer, Float> accumulatedGradientMean,
+            Map<Integer, Float> accumulatedGradientVariance,
+            SGDExtension sgdExtension) {
+        float step = 0.0f;
+        float adaptedLearningRate = 0.0f;
+
+        switch (sgdExtension) {
+            case NONE:
+                step = partial * learningRate;
+                break;
+            case ADAGRAD:
+                if (accumulatedGradientSquares.get(variableIndex) == null) {
+                    accumulatedGradientSquares.put(variableIndex, (float)Math.pow(partial, 2.0f));
+                } else {
+                    accumulatedGradientSquares.put(variableIndex, accumulatedGradientSquares.get(variableIndex)
+                            + (float)Math.pow(partial, 2.0f));
+                }
```

Comment:
For these map patterns, just use a single if (no else) and put in the default value.
In this case, `0.0f`.
Then just do the normal case at the base indentation level (adjacent to the if).